src/core/handle.ml

```diff
@@ -258,50 +258,100 @@ let handle_cmd : sig_state -> p_command -> sig_state * proof_data option =
+        (* Reverse the list of constructors previously computed to preserve
+           the initial order *)
+        let cons_list = List.rev cons_list_rev in
+        (ss, cons_list::cons_list_list)
+      in
+      let (ss, cons_list_list_rev) =
+        List.fold_left add_constructors (ss, []) il
       in
-      let (ss, cons_list_rev) = List.fold_left add_cons (ss, []) l in
-      (* Reverse the list of constructors previously computed to preserve the
-         initial order *)
-      let cons_list = List.rev cons_list_rev  in
-      (* Compute the induction principle *)
-      let rec_typ = Inductive.gen_rec_type ss cmd.pos ind_sym cons_list in
-      (* Check the type of the induction principle *)
-      Inductive.check_type_ind cmd.pos rec_typ;
-      (* Add the induction principle in the signature *)
-      let rec_name = Pos.make cmd.pos (Parser.add_prefix "ind_" id.elt) in
-      if Sign.mem ss.signature rec_name.elt then
-        fatal cmd.pos "Symbol [%s] already exists." rec_name.elt;
-      if StrSet.mem id.elt !(ss.signature.sign_idents) then
-        Sign.add_ident ss.signature rec_name.elt;
-      let (ss, rec_sym) =
-        Sig_state.add_symbol ss e Defin Eager rec_name rec_typ [] None
+      let ind_list = List.combine_rev ind_typ_list_rev cons_list_list_rev in
+      (* STEP 3 - Generate the induction principle. *)
+        (* A - Compute the induction principle *)
+      let rec_typ_list_rev, assoc_predicates =
+        Inductive.gen_rec_type ss cmd.pos ind_list
+      in
+      let check_and_add rec_typ id ss =
+        (* B - Check the type of the induction principle *)
+        Inductive.check_type_ind cmd.pos rec_typ;
+        (* C - Add the induction principle in the signature *)
+        let rec_name =
+          Pos.make cmd.pos (Parser.add_prefix "ind_" id.sym_name)
+        in
+        if Sign.mem ss.signature rec_name.elt then
+          fatal cmd.pos "Symbol [%s] already exists." rec_name.elt;
+        if StrSet.mem id.sym_name !(ss.signature.sign_idents) then
+          Sign.add_ident ss.signature rec_name.elt;
+        let (ss, rec_sym) =
+          Sig_state.add_symbol ss e Defin Eager rec_name rec_typ [] None
+        in
+        if !log_enabled then
+          log_hndl "The induction principle named %a is %a"
+            Pretty.pp_ident rec_name Print.pp_term rec_typ;
+        ss, rec_sym
+      in
+      let ind_list =
+        List.combine3
+          ind_typ_list_rev cons_list_list_rev rec_typ_list_rev
+      in
+      let map_triplet :
+            Sig_state.t -> (sym * sym list * term) list
+            -> Sig_state.t * (sym * sym list * sym) list = fun ss ind_list ->
+        let rec aux ss ind_list acc =
+          match ind_list with
+          | [] -> ss, acc
+          | (i,c,r)::q ->
+              let ss, rec_sym = check_and_add r i ss in
+              aux ss q ((i, c, rec_sym)::acc)
+        in
+        aux ss ind_list []
+      in
+      let ss, ind_list = map_triplet ss ind_list in
```

Comment:
Why not replacing the lines 298 and 323-339 by:
```
let (ss, rec_sym_list) = List.fold_left2 add_recursor (ss, []) ind_typ_list_rev rec_typ_list_rev in
let ind_list = combine_rev3 ind_typ_list_rev cons_list_list_rev (List.rev rec_sym_list) in
```

Reply:
It's a bit ambigous because I changed the previous lines...

Reply:
I don't know how I can do that.
It's really closed to `val fold_left_map : ('a -> 'b -> 'a * 'c) -> 'a -> 'b list -> 'a * 'c list
fold_left_map is a combination of fold_left and map that threads an accumulator through calls to f` 

but that's all.

Reply:
Why the following doesn't work?
```
      let add_recursor (ss, rec_sym_list) ind_sym rec_typ =
        ... (* code of check_and_add with id replaced by ind_sym except last line *)
        (ss, rec_sym::rec_sym_list)
      in
let (ss, rec_sym_list) = List.fold_left2 add_recursor (ss, []) ind_typ_list_rev rec_typ_list_rev in
let ind_list = combine_rev3 ind_typ_list_rev cons_list_list_rev (List.rev rec_sym_list) in
```

Reply:
OK, je n'ai pas les capacités pour être ce genre de choses.
Mais ca marche, merci.